src/frontend/swo/core.ts

```diff
@@ -69,8 +69,10 @@ class ITMDecoder extends EventEmitter {
     }
 
     private rxWriteByte(byte: number): boolean {
-        this.rxBuffer.writeUInt8(byte, this.rxCount);
-        this.rxCount++;
+        if(this.rxCount < this.rxBuffer.length) {
+            this.rxBuffer.writeUInt8(byte, this.rxCount);
+            this.rxCount++;
+        }
```

Comment:
Could you explain what is going on here? Please refresh my memory about the protocol. Maybe you fixed a problem that already existed but I can't tell. If you explain, it will save me a bunch of time researching stuff

Reply:
The main issue was below in the case Status.TIMESTAMP

This technically isn't an issue for a well formed ITM packet stream but I found PeMicro doesn't always produce a  well formed stream.

A "local timestamp" packet is a variable length packet with 1-5 bytes. The end of timestamp is marked by the upper most bit "byte & 0x80".

The issue with the code is that if the stream initiated a Status.TIMESTAMP and the end bit "byte & 0x80" didnt come in the first five bytes, then on the 6th byte this.rxWriteByte(byte); would overflow this.rxBuffer raising an uncaught exception. (this.rxBuffer is length 5 because timestamp is max 5 bytes)

The problem, at least in the case of PeMicro is that during a clock change the stream could start temporarily send garbage. This state machine would get stuck in case Status.TIMESTAMP because it kept overflowing this.rxBuffer even after the stream started sending good data again (because it wasn't timestamp data anymore).

I made it so it counted the number of received bytes in "case Status.TIMESTAMP" (by checking the return of rxWriteByte) and will exit that state if it received more than 5 since it couldnt be a valid timestamp anymore.

https://developer.arm.com/documentation/ddi0403/d/Appendices/Debug-ITM-and-DWT-Packet-Protocol/Packet-descriptions/Local-timestamp-packets?lang=en

Technically I could remove the change in rxWriteByte and let it throw an exception since the change in case Status.TIMESTAMP prevents the overflow as well.

Reply:
It might be a good idea to keep rxWriteByte as it was and only include the change in "case Status.TIMESTAMP". Preventing the overflow in rxWriteByte might mask future problems